[SYCL][Doc] Extended group load/store APIs proposal

aelizaro opened the pull request:
An initial draft of extended group load/store APIs to provide capabilities to work with temporary memory buffers and load/store multiple elements per work item.

aelizaro commented:
@andreyfe1, @Pennycook, could you please take a look at the initial proposal?

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+* link:../experimental/sycl_ext_oneapi_properties.asciidoc[
+  sycl_ext_oneapi_properties].
+
+== Current State
```
I think you should replace this section with one that defines replacements for the single value functions, consistent with your other interfaces, e.g.:

```c++
// Load API
template<typename GroupHelper, typename InputIterator, typename OutputIterator,
        std::size_t ElementsPerWorkItem, typename Properties = properties<>>
void load_over_group(GroupHelper gh, InputIteratorT in_ptr,
                      OutputIteratorT out_ptr, Properties = {});
```

aelizaro replied:
Replaced for single values, please, take a look, not sure if we should take Output& in load case or just return it from function

Pennycook replied:
I'm usually a fan of using return values but in this case, I see multiple advantages to taking the output by reference:
- It's consistent with the `span` overload
- We can deduce the desired output type (which may be useful if this interface later supports type conversion)

Pennycook replied:
Addressed in 349f2592cb7feb996f4590565efa9312ba899773.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+
+* specify data placement type via properties (1) or group_memory_helper (*naming is TBD) class (5)
+
+* extra options such as setting boundry values or limiting numbers of work items can be also specified via group_memory_helper
```
```suggestion
* extra options such as setting boundary values or limiting numbers of work items can be also specified via group_memory_helper
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+
+The extended APIs provides the following features:
+
+* load/store with temporary memory buffer, which can be passed via GroupHelper such as `group_with_scratchpad` link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc#group-helper[group_with_scratchpad description from sort over group proposal]
```
Should there be overloads that _don't_ require a temporary memory buffer?

aelizaro replied:
it can be just the case when instead of `GroupHelper` just a group object is passed, but there won't be performance benefits in this case, I assume

Pennycook replied:
I assumed that the memory provided by the `GroupHelper` would only be used if the data needed to be transposed, and only if the group is a work-group.  It seems like if the data doesn't need to be transposed it can be returned directly to the work-item (without going through intermediate memory), and if the group is a sub-group the transpose could use shuffles.

Pennycook replied:
Resolved; there are now overloads accepting just a `Group`.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+}
+----
+
+(1) _Preconditions_: `in_ptr` must be the same for all work-items
+in the group.
```
Because of this restriction, I think we should rename `load_over_group` to `joint_load` and `store_over_group` to `joint_store`.  This is more aligned with the naming convention of SYCL 2020 group algorithms (e.g. `joint_reduce` requires all work-items to provide the same input range to reduce), and I think it's a more accurate description (because conceptually, the work-items in the group are cooperating to perform a single load/store).

Pennycook replied:
Resolved; these have been renamed.

Pennycook replied:
@aelizaro - Sorry for reopening this comment, but when I was reading back through again, I realized that there might be some ambiguity here.  I think I got confused and steered you in the wrong direction.

The description of "single value" load says:
```c++
template<typename Group, typename InputIteratorT,
         typename OutputT, typename Properties = properties<>>
void joint_load(Group g, InputIteratorT in_ptr, OutputT& out, Properties = {})
```

> _Effects_: Loads single element from `in_ptr` to `out`.

Are all the work-items loading the same (single) value from `in_ptr` (i.e. broadcasting a result from memory)?  Or are all work-items collaborating to load a contiguous block, and returning one element from the block to each work-item?

If the latter, I think the description of _Effects_ needs to be extended to say that a different value is returned to each work-item.  I also think that I was wrong to suggest `joint_load`/`store`; if a different value is returned to each work-item, the naming should be `group_load`/`store`.

aelizaro replied:
the latter, I will rename it back

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+                      GroupMemoryHelper<InputT, ItemsPerWorkItem> memory_helper = {}); (4)
+
+// Properties:
+enum class group_algorithm_data_placement {
```
Bikeshedding: This is a very long name, so we should try and find something shorter.

aelizaro replied:
I also haven't yet figured out if ve need input and output data placement properties or just one-way one (as in Example 2)

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+using the `gh` group helper object. Properties can specify data placement.
+
+
+(2) _Preconditions_: `in_ptr` must be the same for all work-items
+in the group.
```
Should there be a precondition that `out` must be a pointer to the private address space?

To my mind, this restriction is what makes this a "load" and not a "copy".  If `out` can be a pointer to global memory, the function would be able to "load" from global memory and implicitly store the result back to global memory, which would be weird.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+work-group or sub-group and some associated scratch space.
+
+_Effects_: Loads `ItemsPerWorkItem` elements from `in_ptr` to `out`
+using the `gh` group helper object. `GroupMemoryHelper` specifies data placement properties and also can work with extra options such as specifying out-of-boundry value and limited work-items number to work with.
```
Instead of using extra options in the `GroupMemoryHelper`, could we make these two new properties?  Then a developer could use them with the overload that doesn't require a `GroupMemoryHelper`.

aelizaro replied:
OK, removed (2) and (4) and added a TODO for other properties, wanna do some prototyping for it

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+work-group or sub-group and some associated scratch space.
+
+_Effects_: Stores `ItemsPerWorkItem` elements from `in` to `out_ptr`
+using the `gh` group helper object. `GroupMemoryHelper` specifies data placement properties and also can work with extra options such as specifying out-of-boundry value and limited work-items number to work with.
```
```suggestion
using the `gh` group helper object. `GroupMemoryHelper` specifies data placement properties and also can work with extra options such as specifying out-of-boundary value and limited work-items number to work with.
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+
+This extension also depends on the following other SYCL extensions:
+
+* link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc[
+  sycl_ext_oneapi_local_memory].
```
Is this really a dependency?  I see you're using it in one of the examples, but I can't see any part of the API that requires this extension.  I think you can move this to a note in the Usage Examples section.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,220 @@
+* link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc[
+  sycl_ext_oneapi_local_memory].
+
+* link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[
+  sycl_ext_oneapi_group_sort].
```
If we're going to use the `GroupHelper` concept across multiple extensions, it might be nice to pull it out into a separate extension.  It's not immediately clear that this sort extension is a dependency only because of the `GroupHelper` concept.

Pennycook replied:
This was resolved in https://github.com/intel/llvm/pull/7202; we decided to refer to the definition in sycl_ext_oneapi_group_sort in all the extensions that use it.

Pennycook reviewed (changes requested):
Please also wrap things at 80 columns; it makes it much easier to review.

dkhaldi commented:
It will be good to have complete kernels for this extension in tests directory.
Specifically, I would like to see the usage of vec, marray, and simd as the span memory storage.

aelizaro commented:
@dkhaldi, do you mean to add full tests without implementation, to illustrate how it supposes to work with sycl's data types?

dkhaldi commented:
> @dkhaldi, do you mean to add full tests without implementation, to illustrate how it supposes to work with sycl's data types?

Right.

aelizaro commented:
Test cases are added to illustrate how APIs should work.
For sycl::vec case, we can use both span approaches (as it is shown in the test) or have a specialization for a single value case - it will look cleaner.
For sycl::marray it works nicely as its internal representation guarantees contiguous storage of memory, so we can use: 
```
sycl::marray<InputT, items_per_thread> data;
sycl_exp::joint_load(item.get_group(), in.get_pointer(), sycl::span<InputT, items_per_thread>{ data.begin(), data.end() });
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+
+== Overview
+
+This extention defines free functions for load/store operations within work
```
```suggestion
This extension defines free functions for load/store operations within work
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+                      OutputIteratorT out_ptr, Properties = {});
+----
+
+== Funtions with fixed-size arrays
```
```suggestion
== Functions with fixed-size arrays
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+void joint_store(Group gh, InputT in,
+                      OutputIteratorT out_ptr, Properties = {});
+----
+
```
I think we need to add the preconditions etc from below to these functions too.  This should be straightforward, mostly a copy-paste.

aelizaro replied:
@Pennycook, thank you, added. Please, also check again for span-based API, I changed the wording for _Constraints_ section, as the GroupHelper template parameter can be just a group as well, as it is shown in Example 1.

Also, I fixed the single value store function with const InputT& instead of just InputT.

@dkhaldi, @Pennycook, I also added an overload for sycl::vec (new functions (3) and (4)), as in the span-based case it looks a bit ugly: https://github.com/intel/llvm-test-suite/blob/0fd0b218b4e0a4c2b486f3863b0244f54aa0717d/SYCL/GroupLoadStore/group_load_store_span_api.cpp#L167

unlike marray, which is nice with begin() and end(): https://github.com/intel/llvm-test-suite/blob/0fd0b218b4e0a4c2b486f3863b0244f54aa0717d/SYCL/GroupLoadStore/group_load_store_span_api.cpp#L136

please, advise if we should keep it

dkhaldi replied:
@aelizaro, thank you. These are good examples to show this feature works on vec and marray.
- Is there an another document that already talks about construction of span using vec and marray? 
- Does the example you added to llvm-test-suite currently run on CPU and GPU? I did not see XFAIL there.

aelizaro replied:
@dkhaldi, thank you

* For the construction of span using vec and marray I don't know if there are any specific docs, but they both are guaranteed to be in memory sequentially and contiguously: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#memory-layout-and-alignment
* I assume examples shouldn't run now at all, since for this functionality there is no implementation available yet, only API proposal and initial prototypes

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+
+== Funtions with fixed-size arrays
+
+This APIs provides the following features:
```
```suggestion
This API provides the following features:
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+template<typename GroupHelper, typename InputIteratorT, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename Properties = properties<>>
+void joint_load(GroupHelper gh, InputIteratorT in_ptr,
+                sycl::span<OutputT, ItemsPerWorkItem> out, Properties = {}); (1)
```
```suggestion
                sycl::span<OutputT, ElementsPerWorkItem> out, Properties = {}); (1)
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+
+
+(2) _Preconditions_: `out_ptr` must be the same for all work-items
+in the group. `in` is a sycl::span made from a pointer of private memory space.
```
```suggestion
in the group. `in` must be a `sycl::span` made from a pointer to the private memory space.
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+_Constraints_: Only available if `GroupHelper` was created with a
+work-group or sub-group and some associated scratch space.
+
+_Effects_: Stores `ItemsPerWorkItem` elements from `in` span to `out_ptr`
```
```suggestion
_Effects_: Stores `ElementsPerWorkItem` elements from `in` span to `out_ptr`
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,223 @@
+
+== Usage Examples
+
+1.Example shows the simpliest case without local memory usage of blocked load
```
```suggestion
1.Example shows the simplest case without local memory usage of blocked load
```

Pennycook reviewed (changes requested):
Spotted a few minor issues during review of the latest changes, but I think the fixes are straightforward.  Let me know when they're applied and I'll approve.

aelizaro commented:
moved test to a separate PR: https://github.com/intel/llvm-test-suite/pull/1539

andreyfe1 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,293 @@
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = properties<>>
+void group_load(Group g, InputIteratorT in_ptr,
+                      OutputT& out, Properties = {}); (1)
```
Here and below. I'd suggest to write `// (1)` instead of just `(1)`. Since this is C++ code it will look like a comment. Now it looks like `(1)` is a part of the code

aelizaro replied:
done

andreyfe1 commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,293 @@
+
+_Constraints_: Only available if `Group` is a work-group or sub-group.
+
+_Effects_: Loads single element from `in_ptr` to `out`.
```
Add "using the g group object" or something like that because we don't specify how to use `g` here.

aelizaro replied:
Good point! Actually, I noticed an inconsistency in my understanding here with [the previous proposal](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_group_load_store.asciidoc), @andreyfe1, @Pennycook, please, advice:

I thought that for single value API the following test-case should work ok: https://github.com/intel/llvm-test-suite/blob/8950ae921bd2be785d42de113a278f91571d830e/SYCL/GroupLoadStore/group_load_store_api.cpp#L19
and we don't need to pass different pointers for different groups and use addresses like: `ptr + g.get_group_id(0) * g.get_local_range(0)  + g.get_local_id(0)`.

But if it should work as `ptr + get_local_id()`, then if we want to load from global memory for each work item, it will look like this:

```
      cgh.parallel_for(sycl::nd_range<1>(size, block_size),
                       [=](sycl::nd_item<1> item) {
                         auto g = item.get_group();

                         InputT data;

                         sycl_exp::group_load(g, in.get_pointer() +  g.get_group_id(0) * g.get_local_range(0), data);

                         data += item.get_global_linear_id() * 100000;

                         sycl_exp::group_store(g, data, out.get_pointer() + g.get_group_id(0) * g.get_local_range(0));
                       });
    });
```
or is it just not the primary use case we want to cover? As we assume the same behaviour as mentioned above for sycl::vec and span-based API:

```
      cgh.parallel_for(
          sycl::nd_range<1>(block_count * block_size, block_size),
          [=](sycl::nd_item<1> item) {
            InputT data[items_per_thread];

            sycl_exp::group_load(item.get_group(), in.get_pointer(),
                                 sycl::span{data});

            for (int i = 0; i < items_per_thread; ++i) {
              data[i] += item.get_global_linear_id() * 100000 + i * 1000;
            }

            sycl_exp::group_store(item.get_group(), sycl::span{data},
                                  out.get_pointer());
          });
```
It will load a contiguous block of memory from the single global pointer by default:
|         wi[0]         |             wi[1]    |          wi[2]        |
|-----------------|------------------|-----------------|
| in[0] in[1] in[2] | in[3] in[4] in[5] | in[6] in[7] in[8] |

in striped case:
|         wi[0]         |             wi[1]    |          wi[2]        |
|-----------------|------------------|-----------------|
| in[0] in[3] in[6] | in[1] in[4] in[7] | in[2] in[5] in[8] |

aelizaro replied:
or it is still a matter of naming and we should distinguish group_load/store, which works with local_id() in the group and block_load/store which works with global_id()

Pennycook replied:
I think it's important that the group algorithms assume that only the group passed into the function is involved in the call.  If the behavior was defined such that `sycl::group_load(g, ptr)` looked at the _global_ work-item ID, then the only way to access all the data in `ptr` would be to ensure that all the groups execute the `load` function.  I think most developers write code from the perspective of a single work-item, and so they're used to having to calculate things like per-group offsets.

The use-case that you mentioned is important, though, and the planned solution is to provide different types of group (e.g. see [sycl_ext_oneapi_root_group](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_root_group.asciidoc)):

```c++
sycl::root_group = it.get_root_group();
sycl::group_load(root_group, in.get_pointer(), sycl::span{data});
```

The `root_group` is a group containing all of the work-items in the kernel, and so calling `root_group::local_id` is equivalent to calling `group::global_id`.

aelizaro replied:
ooh, I see, that is a cool proposal. I will align with the initial one, add a detailed description for the expected behaviour and update the tests, thank you!

aelizaro replied:
both the proposal and test cases are updated to work with the local scope

aelizaro commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,293 @@
+_Constraints_: Only available if `Group` is a work-group or sub-group.
+
+_Effects_: Loads single element from `in_ptr` to `out`.
+Properties TBD (setting boundary values or limiting numbers of work items).
```
I think it still makes sense to keep it, as we won't be able to add it as an extra API with default parameter in the future without breaking compatibility as there will be ambiguity

AlexeySachkov replied:
Can we then add a more user-friendly phrase like: this argument is reserved for future revisions of this extension and it is ignored for now. Or do we want to say that passing any properties here results in a compilation error or UB?

aelizaro replied:
updates, please, check wording

AlexeySachkov commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,299 @@
+
+_Constraints_: Only available if `Group` is a work-group or sub-group.
+
+_Effects_: Loads single element from `in_ptr` to `out` by using the `g` group
+object to identify memory location. For the 1d case it would be:
+`g.get_group_id(0) * g.get_local_range(0)  + g.get_local_id(0)`
```
Can we specify how exactly group `g` object identifies memory location which needs to be loaded? Based on the example, I _assume_ that every work item loads an element address of which is calculated as `in_ptr + g.get_local_linear_id()` - is it correct understanding?

It is just I don't think that it is enough for the spec to provide an example instead of actually documenting how it is supposed to work.

aelizaro replied:
yes, you are right, updated description everywhere

AlexeySachkov commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,299 @@
+(1) _Preconditions_: `in_ptr` must be the same for all work-items
+in the group.
+
+_Constraints_: Only available if `Group` is a work-group or sub-group.
```
I think we may want to add some additional constraints on `InputIteratorT` and `OutputT` here. 

For example, do we require `InputIteratorT` to be [random access iterator](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator) or simple [forward iterator](https://en.cppreference.com/w/cpp/named_req/ForwardIterator) should be enough?

Do we require type `T` which is obtained by dereferencing iterator of type `InputIteratorT` to be exactly the same as `OutputT`? Or is it enough if `T` is convertible to `OutputT`?

The same comment applies to all other functions and overloads proposed in this extension.

aelizaro replied:
Updated, please, check if the wording is ok

AlexeySachkov commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,299 @@
+_Constraints_: Only available if `Group` is a work-group or sub-group.
+
+_Effects_: Loads `N` elements from `in_ptr` to `out`
+using the `g` group object. Properties can specify data placement.
```
What is the default placement if no properties provided? Do we need to add a comment describing potential future direction, i.e. how exactly we imagine properties affect data placement?

aelizaro replied:
Specified default one

AlexeySachkov commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,299 @@
+}
+----
+
+(6) _Preconditions_: `in_ptr` must be the same for all work-items
```
It seems like for this overload we actually do have properties we want to support: do we want to restrict the API to them? Like specifying which exact properties are supported by certain overloads? And what happens if wrong properties are passed?

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+
+// Load API scalar
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = properties<>>
```
Properties extension doesn't provide `properties<>`, IIUC
```suggestion
         typename OutputT, typename Properties = decltype(properties())>
```

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+|{2, 5, 8, 11}
+|===
+
+There is a property that satisfies
```
I'm not a native speaker, so might be wrong here. That said, my first impression was that one of other extensions has already defined that property, but it seems not to be so (because group sort has separate input/output properties for that, albeit using the same wording as here). 

Can we change it to something along the lines of "this extension adds new property(ies)..."?

aelizaro replied:
Good point, changed description and added references to group sort's input/output_data_placement

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+_Effects_: Loads single element from `in_ptr` to `out` by using the `g` group
+object to identify memory location as `in_ptr` + `g.get_local_linear_id()`.
+
+`Properties` argument is reserved for future revisions of this extention and is
+ignored now. Default value is empty `sycl::ext::oneapi::experimental::properties<>`
+May be used in future for setting boundary values or limiting numbers of work items.
```
I'd suggest to introduce `alignment` property because otherwise we won't be able to optimize out a runtime alignment check.

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+`out_ptr[gh.get_local_linear_id() + gh.get_local_linear_range() * i]` = `in[i];`
+for `i` between `0` and `ItemsPerWorkItem`.
+
+(8)_Effects_: Returns size of temporary memory (in bytes) that is required for
+scratch space in `GroupHelper`. Result depends on type `T`, `ElementsPerWorkItem`
+and the scope parameter: use sycl::memory_scope::work_group to get memory size
+required for each work-group; use sycl::memory_scope::sub_group to get memory
+size required for each sub-group. If other scope values are passed, behavior is
+unspecified.
```
```
constexpr std::size_t memory_required(sycl::memory_scope scope,
                                      std::size_t block_size); // (8)
```

`block_size` parameter isn't described here.

aelizaro replied:
Added description

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+
+q.submit([&](sycl::handler& cgh) {
+    constexpr auto temp_memory_size = sycl_exp::memory_required<T, items_per_thread>(
+        sycl::memory_scope::work_group, block_size);
+    sycl::local_accessor<std::byte> buf(temp_memory_size, cgh);
+    cgh.parallel_for(
+        sycl::nd_range<1>(global_size, local_size),
```
Should there be some relation between `block_size` and `local_size`?

aelizaro replied:
fixed it in this example, should be same for work-group case, for sub-group case I assume it should be something like local_size = block_size * number of subgroups, but I am not sure performance-wise

aelovikov-intel commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,413 @@
+(8)_Effects_: Returns size of temporary memory (in bytes) that is required for
+scratch space in `GroupHelper`. Result depends on type `T`, `ElementsPerWorkItem`
+and the scope parameter: use sycl::memory_scope::work_group to get memory size
+required for each work-group; use sycl::memory_scope::sub_group to get memory
+size required for each sub-group. If other scope values are passed, behavior is
```
> to get memory size required for each sub-group

Would user have to allocate `returned_value * number_of_sub_groups` bytes? If so, that would force usage of `[[intel::reqd_sub_group_size]]`` attribute, because otherwise that number won't be computable.

aelizaro replied:
Does group_with_scratchpad works only for sycl::group or sub_group also? If the former, then I need to double-check, maybe it is not even a use-case for sub_group to use this extra memory

aelovikov-intel replied:
@Pennycook , do you know off the top of your head?

Pennycook replied:
I believe `group_with_scratchpad` supports any type of group.  My understanding is that it's just a way to represent a group object and a pointer simultaneously -- there shouldn't be any requirement that the pointer is in any particular memory, so a group algorithm with sub-group scope could use local or global memory.

I agree it would be easier to reason about how many sub-groups exist for kernels with `reqd_sub_group_size`.

bader reviewed (commented):
@aelizaro, when do you expect to have the final version of spec? I'm interested in implementing sub-group operations with cache hints, so I can help with addressing comments if needed.

aelizaro commented:
> @aelizaro, when do you expect to have the final version of spec? I'm interested in implementing sub-group operations with cache hints, so I can help with addressing comments if needed.

@bader, terribly sorry, I missed your comments completely. Ideally it will be great to finalize current version as soon as possible. so we won't lose it for a long time again. I am currently working to address your comments.

aelizaro commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,415 @@
+ignored now. Default value is empty `sycl::ext::oneapi::experimental::properties<>`
+May be used in future for setting boundary values or limiting numbers of work items.
+
+== `sycl::vec` Functions
```
`sycl::marray` has iterators, so it can use sycl::span API, as:

`
sycl::marray<T, items_per_thread> marray;
sycl::span<T, items_per_thread>{ marray.begin(), marray.end() };`

Do you think it needs an extra overload? (for sycl::vec it is a bit more ugly, that is why we added an extra overload)

bader commented:
> > @aelizaro, when do you expect to have the final version of spec? I'm interested in implementing sub-group operations with cache hints, so I can help with addressing comments if needed.
> 
> @bader, terribly sorry, I missed your comments completely. Ideally it will be great to finalize current version as soon as possible. so we won't lose it for a long time again. I am currently working to address your comments.

Thanks. @Pennycook, could you approve this PR if your comments are addressed?

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,469 @@
+This extension also depends on the following other SYCL extensions:
+
+* link:../proposed/sycl_ext_oneapi_group_sort.asciidoc[
+  sycl_ext_oneapi_group_sort] (`group_with_scratchpad` class used as GroupHelper).
```
I think this dependency only exists for the "Fixed-size array Overload" parts of the API.  Those parts of the extension also seem to have some unresolved issues in the specification.  Do we think these parts are important enough to do right away, or could they be moved to a "Future direction" section in the document?

It would be nice to implement the other parts of this spec soon, so that we can deprecate / remove the `sub_group::load` and `sub_group::store` functions from the DPC++ implementation.  I was wondering if we could implement these other parts sooner if we delay the "Fixed-size array Overload" parts until later.

Pennycook replied:
I think this is a good suggestion.

The single-value load/store functions are the most important for existing SYCL codes, and even just supporting these new functions would expose important new functionality for our GPUs (because the `sub_group::` versions don't allow selection of blocked/striped layout).

The fixed-size array overload parts are mainly useful to support patterns found in CUDA kernels that employ CUB.  They're important too, but could be implemented in a Phase 2.

aelizaro replied:
Agree, I moved everything related to group_with_scratchpad into Future direction section. @gmlueck, @Pennycook please take a look if it looks in place now

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,469 @@
+}
+----
+
+(1) _Preconditions_: `in_iter` must be the same for all work-items
```
These numeric callouts (e.g. `(1)`) seem to add more confusion.  For example, why does _Preconditions_ have the (1) while _Constraints_ does not?  I'd suggest removing them.

I usually only add the callouts if there are multiple functions in the same code block and the _Constraints_, etc. are different for each function.  In your case, the code block only has one function, so the numeric callout doesn't seem necessary.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,469 @@
+
+// Load scalar
+template<typename Group, typename InputIteratorT,
+         typename OutputT, typename Properties = decltype(properties())>
```
```suggestion
         typename OutputT, typename Properties = empty_properties_t>
```

And other places too.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,505 @@
+namespace sycl::ext::oneapi::experimental {
+
+// Load API
+template<typename GroupHelper, typename InputIteratorT, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename Properties = empty_properties_t>
+void group_load(GroupHelper gh, InputIteratorT in_iter,
```
Should this be a `Group` instead?

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,505 @@
+namespace sycl::ext::oneapi::experimental {
+
+// Store API
+template<typename GroupHelper, typename OutputIteratorT, typename InputT,
+        std::size_t ElementsPerWorkItem, typename Properties = empty_properties_t>
+void group_store(GroupHelper gh, sycl::span<InputT, ItemsPerWorkItem> in,
+                 OutputIteratorT out_iter,  Properties = {});
```
Same comment as above -- should this be a `Group` now?

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+// Load API
+template<typename Group, typename InputIteratorT, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename Properties = empty_properties_t>
+void group_load(Group gh, InputIteratorT in_iter,
```
```suggestion
void group_load(Group g, InputIteratorT in_iter,
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+* `InputIteratorT` suppose to be a random access iterator.
+* Value type of `InputIteratorT` must be convertible to `OutputT`.
+
+_Effects_: Loads `ElementsPerWorkItem` elements from `in_iter` to `out`
+using the `gh` group helper object.
+Properties may specify xref:data_placement[data placement].
+Default placement is a blocked one:
+`out[i]` = `in_iter[gh.get_local_linear_id() * ElementsPerWorkItem + i];`
+in striped case:
+`out[i]` = `in_iter[gh.get_local_linear_id() + gh.get_local_linear_range() * i];`
+for `i` between `0` and `ElementsPerWorkItem`.
```
```suggestion
_Effects_: Loads `ElementsPerWorkItem` elements from `in_iter` to `out`
using the `g` group object.
Properties may specify xref:data_placement[data placement].
Default placement is a blocked one:
`out[i]` = `in_iter[g.get_local_linear_id() * ElementsPerWorkItem + i];`
in striped case:
`out[i]` = `in_iter[g.get_local_linear_id() + g.get_local_linear_range() * i];`
for `i` between `0` and `ElementsPerWorkItem`.
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+// Store API
+template<typename Group, typename OutputIteratorT, typename InputT,
+        std::size_t ElementsPerWorkItem, typename Properties = empty_properties_t>
+void group_store(Group gh, sycl::span<InputT, ItemsPerWorkItem> in,
```
```suggestion
void group_store(Group g, sycl::span<InputT, ItemsPerWorkItem> in,
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+* `OutputIteratorT` suppose to be a random access iterator.
+* `InputT` must be convertible to value type of `OutputIteratorT`.
+
+_Effects_: Stores `ElementsPerWorkItem` elements from `in` span to `out_iter`
+using the `gh` group helper object.
+
+Properties may specify xref:data_placement[data placement].
+Default placement is a blocked one:
+`out_iter[gh.get_local_linear_id() * ItemsPerWorkItem + i]` = `in[i];`
+in striped case:
+`out_iter[gh.get_local_linear_id() + gh.get_local_linear_range() * i]` = `in[i];`
+for `i` between `0` and `ItemsPerWorkItem`.
```
```suggestion
_Effects_: Stores `ElementsPerWorkItem` elements from `in` span to `out_iter`
using the `g` group object.

Properties may specify xref:data_placement[data placement].
Default placement is a blocked one:
`out_iter[g.get_local_linear_id() * ItemsPerWorkItem + i]` = `in[i];`
in striped case:
`out_iter[g.get_local_linear_id() + gh.get_local_linear_range() * i]` = `in[i];`
for `i` between `0` and `ItemsPerWorkItem`.
```

Pennycook reviewed (commented):
Sorry I didn't spot this yesterday, but now that you've changed `GroupHelper` to `Group` I think we should also change the names from `gh` to `g` for consistency, and remove a few references to "group helper" in the early sections.

aelizaro commented:
> Sorry I didn't spot this yesterday, but now that you've changed `GroupHelper` to `Group` I think we should also change the names from `gh` to `g` for consistency, and remove a few references to "group helper" in the early sections.

Sorry, my bad, I didn't think about wording, updated.

Pennycook reviewed (approved):
My concerns are addressed.  But @intel/llvm-gatekeepers, please wait until there is also an approval from @gmlueck.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+
+To specify a correct data placement for placing of resulting data
+there is a enum
+(proposed in link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[`group_sort extention`]):
```
```suggestion
(proposed also in link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc[`group_sort extention`]):
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+    template<group_algorithm_data_placement type>
+    struct data_placement;
+}
+----
```
This is not the right way to define a compile-time property.  It should be like this instead:

```
namespace sycl::ext::oneapi::experimental {

struct data_placement_key {
  template <group_algorithm_data_placement Type>
  using value_t =
      property_value<data_placement_key,
                     std::integral_constant<group_algorithm_data_placement, Type>>;
};

template <group_algorithm_data_placement Type>
inline constexpr data_placement_key::value_t<Type> data_placement;

inline constexpr data_placement_key::value_t<group_algorithm_data_placement::blocked> data_placement_blocked;
inline constexpr data_placement_key::value_t<group_algorithm_data_placement::striped> data_placement_striped;

} // namespace sycl::ext::oneapi::experimental
```

An example usage would then be like:

```
group_load(g, input, output_span, data_placement_blocked);
```

BTW, I agree with the previous comment that the name `group_algorithm_data_placement` is too long.  How about `data_placement_enum`.  We have a convention with other properties where we use the `_enum` suffix.

aelizaro replied:
Applied suggestion, I assume it is also need to be in experimental::property namespace?
And another thought: should data_placement also be a namespace then or is it too much?

The previous version was based on group_sort proposal, I assume it would also require similar update

aelizaro replied:
ok, got it, removed property:: namespace, thank you!

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+Example:
+`group_load(g, input, output_span, properties<data_placement<blocked>>{});`
+
+=== Usage Example
```
```suggestion
== Usage Example
```

The examples should be a peer section of "Specification", not a subsection within it.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+sycl::queue q;
+sycl::usm_allocator<T, sycl::usm::alloc::device, 64> allocator(q);
+std::vector<T, decltype(allocator)> input(size, allocator);
+std::vector<T, decltype(allocator)> output(size, allocator);
```
We don't support `usm_allocator` with `usm::alloc::device` memory.  This example also captures `std::vector` as a kernel argument which is not officially supported as an argument type.  I think it would be simpler to just use `malloc_device` here.

aelizaro replied:
Fixed, thank you

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+
+            // Work with data...
+
+            sycl_exp::group_store(item.get_group(), output.data(), sycl::span{ data });
```
Aren't all work-items storing to the same 4 elements of `output`?

aelizaro replied:
Good catch! I think it was here since this logic supposed to be internal, but in current version, I added offset

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+});
+----
+
+=== Future Directions
```
```suggestion
== Future Directions
```

Sections below should also be reduced by one level.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
I think we should name this file "proposed/sycl_ext_oneapi_group_load_store.asciidoc".  Once we implement this extension and move it to "experimental", we can move the old one to "deprecated/sycl_ext_oneapi_group_load_store.asciidoc".

aelizaro replied:
Thank you, renamed

gmlueck reviewed (commented):
This is looking good!  Just a few comments below.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+
+._Preconditions_:
+* `in_iter` must be the same for all work-items in the group.
+* `out` must be a `sycl::span` made from a pointer to the private memory space.
```
This is a good catch. I think we should probably clarify that the other interfaces are required to be allocated in the private address space too.

The original thinking here was that these functions are intended to be used to load from/store to memory.  If the arguments could be in the local or global memory space, the functionality would be closer to a "copy" operation.

gmlueck replied:
Is there really any reason to limit it to the private address space, though?  If the implementation works for any address space, why add a limitation?

With the existing API, we could do something like this:

```
acc[i] = sg.load(ptr);
```

The most natural way to write this in the new API would be:

```
group_load(sg, ptr, acc[i]);
```

However, that would violate the precondition because `acc[i]` is not in the private address space.  Instead, a conformant application would need to write:

```
T x;
group_load(sg, ptr, x);
acc[i] = x;
```

Wouldn't it be easier to just allow the destination of the load to be in any address space?

aelizaro replied:
I don't have a strong opinion here from API point of view, but I assume for performance value it makes sense to guarantee that it is private memory, as I am not sure, that we can check it inside of the implementation.
@aelovikov-intel, could you please, advise here?

gmlueck replied:
The implementation could be written to ensure that the SPIR-V instruction does always access private memory.  For example (pseudo-code):

```
void group_load(Group g, InputIteratorT in_iter, OutputT& out) {
  OutputT tmp = __spirv_SubgroupBlockReadINTEL(in_iter);
  out = tmp;
}
```

I assume the optimizer will eliminate the use of `tmp`.

Pennycook replied:
> ```
> acc[i] = sg.load(ptr);
> ```

This is doing two things:

1. `sg.load(ptr)` loads from `ptr` into a variable (in private memory) and returns it.
2. `=` assigns the value from private memory to `acc[i]`

> ```
> T x;
> group_load(sg, ptr, x);
> acc[i] = x;
> ```

This is definitely more verbose, but I like that it clearly remains as two steps.  Somebody might otherwise expect `group_load(sg, ptr, &acc[i])` to issue something like a DMA to "load" directly into the memory associated with `acc[i]`.


> Wouldn't it be easier to just allow the destination of the load to be in any address space?

If we did this, I think we'd want to say in the specification that if the pointer is in the local or global address space, the implementation will/can load into private memory first.

But I'm still not sure this is a good idea.  Consider what this looks like for the store version of things:

```c++
group_store(g, &acc[i], ptr);
```

...would people expect that to load the value from `acc[i]` into private memory, then write it out to `ptr`?  Compare with:

```c++
T x = acc[i];
group_store(g, &x, ptr);
```

aelovikov-intel replied:
> The implementation could be written to ensure that the SPIR-V instruction does always access private memory. For example (pseudo-code)

It has to be written like that to satisfy the implicit conversion to the output type that the extension mandates (or at least used to in previous revisions).

gmlueck replied:
> to satisfy the implicit conversion to the output type that the extension mandates (or at least used to in previous revisions)

The language about conversion is still in the spec.

gmlueck replied:
The reason I worry about adding the precondition about the private address space is that I think it's easy to miss.  The compiler won't diagnose an error, and I think our implementation will run correctly even if you pass a reference to a different address space.  I think this will lead to a lot of code that is non-conforming.  If our implementation ever changes to actually require the private address space, this code will suddenly break.

If we really want to require the private address space, I think we should change the parameter to be a `multi_ptr` with `address_space::private_space`.  That will force people to think about the address space, and the compiler will enforce it.  However, this also means that the common code pattern where you load into a block-scope variable will become much more verbose.

Given @aelovikov-intel's comment, it seems like we need the internal temporary variable anyway, so it seems like the implementation will end up allowing the parameter to be a reference to any address space anyways.

Pennycook replied:
I think the behavior you described is exactly why this ended up as a precondition.  It's UB, so anything can happen (including that it works).  We can't prove that it's a private pointer in the general case (e.g., for generic pointers).  But I agree that this is going to result in lots of bugs that are hard to track down.

I really don't want developers to have to use a `multi_ptr` to get good performance here, so you've convinced me that removing the restriction is a better solution.

How do you feel about my suggestion that we document that implementations can/will stage things via private memory?  I think we're going eventually going to end up with a `group_copy` function too, and I want to make sure that developers understand the difference.

aelizaro replied:
ok, I removed private mentioning from preconditions, thank you!

gmlueck replied:
> How do you feel about my suggestion that we document that implementations can/will stage things via private memory?

This seems like an implementation detail to me.

> I think we're eventually going to end up with a `group_copy` function too, and I want to make sure that developers understand the difference.

How would `group_copy` functions be different?  In particular, the "Fixed-size Array Overload" from this spec seems like it is copying memory from an iterator to another container.

Pennycook replied:
> How would `group_copy` functions be different? In particular, the "Fixed-size Array Overload" from this spec seems like it is copying memory from an iterator to another container.

I think the difference is mainly intent: the developer can be clearer about what they're trying to express, and the implementation can have different behavior that's aligned with the developer's intent.

`group_load` was intended to mean "load this value from memory, in the most efficient way possible".  Developers should expect it to act as a (potentially faster) drop-in replacement for something like running `T x = *ptr;`, on each work-item.  The reason the function exists is to help implementations reason about cases where all the work-items in the group are accessing `ptr` in a pattern that can be optimized.  Conceptually, the developer describes how all the accesses made by individual work-items can be combined into a larger block.  Each work-item owns a dedicated piece of the output, but collaborating with other work-items improves performance.

`group_copy` will mean "copy these values from memory X to memory Y, in the most efficient way possible".  Developers should expect it to act as a (potentially faster) drop-in replacement for something like `memcpy` on a single work-item.  Conceptually, the developer describes an operation that needs to happen (i.e., the copy), and the implementation gets to decide how that operation is distributed across work-items or whether it can employ dedicated hardware (e.g. DMAs).  Work-items don't own dedicated pieces of the output, but collaborate to move all the bytes from X to Y.

gmlueck replied:
Yes, I see the difference now.  With the load / store functions, each work-item provides its contribution to the values that are loaded or stored via a non-uniform parameter.  With the copy function, there would just be two uniform iterators, and it would be undefined which work-items copied which elements.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,522 @@
+._Constraints_:
+* Only available if `Group` is a work-group or sub-group.
+* `InputIteratorT` suppose to be a random access iterator.
+* Value type of `InputIteratorT` must be convertible to `OutputT`.
```
I was thinking about the implementation a bit.  I presume these functions will use **OpSubgroupBlockReadINTEL** and **OpSubgroupBlockWriteINTEL** from [SPV_INTEL_subgroups](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_subgroups.html).

When the type is not a fundamental scalar (e.g. a class), I presume you will just do a bunch of those operations to load or store the object piecemeal.  However, this assumes that the type is trivially copyable.  Do we need to add a constraint like:

> The value type of `InputIteratorT` and the type `OutputT` must both be trivially copyable.

aelovikov-intel replied:
I think being default-constructible would also help. So that I could do

```
using value_type = std::iterator_traits<InputIteratorT>::value_type;
static_assert(sizeof(value_type) == 6); // For weird types.
value_type tmp[NElems];
uintN x = /*spirv-block-load*/;
ushortN y = /*spirv-block-load*/;

memcpy(tmp, &x, ...);
memcpy(tmp + /* ... */, &y, ...);
```

But we can update the spec wording once we have implementation PR that makes use of that.

aelovikov-intel replied:
Or maybe I could do

```
char storage[size_in_bytes];
auto tmp_ptr = reinterpret_cast<value_type*>(storage);
```

and it won't be necessary to default-construct...

gmlueck replied:
You should not do the `reinterpret_cast` like this because this is undefined behavior by C++ rules.  In fact, this code is likely to break if the default constructor is not trivial.  If we need the value type of `InputIteratorT` to be default constructible, we should add that as a constraint.

gmlueck replied:
@aelovikov-intel and I chatted about this.  There are two ways he can write the algorithm, which have slight different requirements on the type:

1. He can assume that the type is "implicit lifetime", or
2. He can assume that the type is both "trivially copyable" and "default constructible".

The two options differ in the case when the default constructor is not trivial but some other constructor (like the copy constructor) is trivial.  We think that either implementation will be performant for types that don't do weird things in their constructors.

The algorithm in option 2 will construct a temporary object and then memcpy data over the object.  The default constructor for some types (e.g. `std::complex`) value-initializes its fields.  We assume that the optimizer will remove this initialization code since the contents of the object will be immediately overwritten with the memcpy, but Andrei will verify.

It seems to me that the constraints in option 2 are easier to explain to users, and it's also nice that there are C++17 traits that correspond to these constraints.  (The trait `std::is_implicit_lifetime` isn't added until C++23.)  Therefore, I think we should go with option 2, and thus we should add a new constraint to these functions saying that the type must be default constructible.

Did I get that right, @aelovikov-intel?

One thing we need clarification on, Andrei -- is this a constraint on `OutputT` or on the value type of `InputIteratorT`?

aelovikov-intel replied:
> Did I get that right, @aelovikov-intel?

Yes! Also, implicit liftime approach is best implementable using C++23's `start_lifetime_as` which is another argument in favor of `default_constructible` trait.

> One thing we need clarification on, Andrei -- is this a constraint on OutputT or on the value type of InputIteratorT?

I think it's both `InputIteratorT` and `OutputIteratorT` for stores:

```
  bytes = block_load
  InputIteratorT::value_type tmp[N]; // default_constructible
  memcpy // trivially_copyable is used here and implicitly in block_load
  out[i] = tmp[i] for ...
```

For store that would be

```
  OutputIteratorT tmp[N]; // default_constructible
  tmp[i] = in[i] for ...
  block_store  // trivially_copyable is implicitly used here
```

aelizaro replied:
Great, added! Am I right what we don't need it for a scalar version (or should I still add it for consistency)?

aelovikov-intel replied:
We have "scalar" block load in SPIR-V, so I expect that it is needed.

aelizaro replied:
added everywhere + also for both types (input/output)

aelovikov-intel commented:
Naive implementation: https://github.com/intel/llvm/pull/13043. I want to implement SPIRV block loads/stores in a separate PR to setup some performance measurements, especially for types with weird sizes.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,553 @@
+
+anchor:optimization_hints[]
+
+As `InputIteratorT` and `OutputIteratorT` are supposed to be random access but
+not garanteed to be contiguous, the following property is introduced to be used
```
```suggestion
As `InputIteratorT` and `OutputIteratorT` are permitted to be random access iterators they
are not guaranteed to be contiguous. The following property is introduced to be used
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,553 @@
+} // namespace sycl::ext::oneapi::experimental
+----
+
+Example, we imply that `input` is contiguous iterator:
+`group_load(g, input, output_span, contiguous_memory);`
```
I think "assert" is stronger here; we're not just implying something, but telling the compiler it's true.  Because of that, we also need to explain what happens if the optimization hint is provided and it's untrue.

```suggestion
For example, we can assert that `input` is a contiguous iterator:
`group_load(g, input, output_span, contiguous_memory);`

If `input` isn't a contiguous iterator, the behavior is undefined.
```

aelizaro commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc:
```diff
@@ -0,0 +1,506 @@
+
+._Preconditions_:
+* `out_iter` must be the same for all work-items in the group.
+* `in` must be a `sycl::span` made from a pointer to the private memory space.
```
removed, thanks

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -391,10 +392,42 @@ For example, we can assert that `input` is a contiguous iterator:
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct full_work_group_hint {
+  using value_t =
+      property_value<full_work_group_hint>;
+};
+
+inline constexpr full_work_group_hint::value_t full_work_group;
+
+struct full_sub_group_hint {
+  using value_t =
+      property_value<full_sub_group_hint>;
+};
+
+inline constexpr full_sub_group_hint::value_t full_sub_group;
```
I think we should combine these into one hint.  We can call it `full_group` for now.

We already know whether we're using a group or a sub-group from the first argument.  There are also going to be other group types in the future, and it would be convenient to be able to re-use the same property across all of those group types as well.

aelovikov-intel replied:
I'm not sure `full_group` is a good name. `parallel_for` requires that global size is a multiple of a work group size, so WG is always "full" in some sense. It might still be that the last SG of each WG isn't full.

Pennycook replied:
Right, but that just means that `full_group` doesn't do anything when the algorithm is invoked with a `sycl::group`, but improves performance when the algorithm is invoked with a `sycl::sub_group`.  The "group" in "full group" is intended to mean "anything satisfying the group concept" and not "`sycl::group`".

I want to push for a renaming of `sycl::group` to `sycl::work_group` to avoid this sort of confusion.

aelovikov-intel replied:
Then we need another property for the WG case, I think.

Pennycook replied:
Why?  Maybe I'm misunderstanding what you're expecting to happen here.  I thought your earlier message was saying that work-groups were always "full"?

aelovikov-intel replied:
They are, but the block-read is SG-level intrinsic. We need SGs to be "full" to use them.

Pennycook replied:
I think we're talking past each other.

Consider the below:

```c++
auto sg = it.get_sub_group();
group_load(sg, input, output_span, properties{contiguous_memory, full_group});
```

The interpretation here is that `sg` represents a full sub-group.

aelovikov-intel replied:
If the user writes

```
  auto g = it.get_group();
  group_load(g, input, output_span, properties{contiguous_memory, full_group});
```

we still can't use block loads without runtime checks. WG is full, but its last SG might not be. Also, `full_group` is meaningless for WG because it always full per SYCL spec `parallel_for`/`nd_range` restrictions.

Pennycook replied:
We couldn't use block loads in this case anyway, because the stride in this case is the size of the work-group and not the sub-group.  You're right that we could generate block loads in some cases (like a work-group + blocked data placement) but I think we'd still need a runtime check because different devices carve up work-groups into sub-groups differently.

`full_group` is currently meaningless for work-groups, but I think it makes more sense to have a generic `full_group` than to make the property sub-group specific.  That allows a developer to write generic code that uses `full_group` and have it work in both cases, and it may be meaningful for future group types.

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -391,10 +392,42 @@ For example, we can assert that `input` is a contiguous iterator:
+==== Groups partitioning
+
+The following properties  is introduced to be used
+as a hint that implementation can use get_max_local_range():
```
Users might not understand why this is helpful for implementations, so I think we should be a little more explicit here.  Maybe something like:

```suggestion
The following property can be used as a hint that
`get_local_range()` is equivalent to `get_max_local_range()`,
which may enable more aggressive optimizations for some
implementations.

[NOTE]
====
Using `full_group` is necessary to generate SPIR-V block read
and block write instructions, because these instructions are
defined to use the maximum group size as the stride.
====
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,588 @@
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct contiguous_memory_hint {
```
```suggestion
struct contiguous_memory_key {
```

The naming convention for properties is that the struct has the suffix "_key" and the name of the property is the same but without the suffix.  Therefore, either name this `contiguous_memory_hint_key` / `contiguous_memory_hint` or  `contiguous_memory_key` / `contiguous_memory`,

As @Pennycook mentioned elsewhere, this is an assertion, not a hint, so I think we should not use "hint" in the name.  Therefore, if you prefer the longer name, it should be like `contiguous_memory_assertion_key`.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,588 @@
+            auto offset = g.get_group_id(0) * g.get_local_range(0) *
+                                       items_per_thread;
+
+            sycl_exp::group_load(g, input + offset, sycl::span{ data }, sycl::contiguous_memory);
```
```suggestion
            sycl_exp::group_load(g, input + offset, sycl::span{ data }, sycl_exp::contiguous_memory);
```

Elsewhere also.

aelizaro commented:
@gmlueck, @Pennycook, @aelovikov-intel, could you please, double-check the PR? (and huge thanks to you for contribution!)

I think I addressed everything except one ongoing discussion about `full_group` hint naming https://github.com/intel/llvm/pull/7593#discussion_r1538044965

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,602 @@
+_Effects_: Loads single element from `in_iter` to `out` by using the `g` group
+object to identify memory location as `in_iter` + `g.get_local_linear_id()`.
+
+`Properties` argument is reserved for future revisions of this extention and is
+ignored now. Default value is empty
+`sycl::ext::oneapi::experimental::empty_properties_t`.
+May be used in future for setting boundary values or limiting numbers of work
+items.
```
Is this still true, or does the `contiguous` property make sense for this overload now?  If it does, I think you should include a similar note as the other overload:

```suggestion
Properties may provide a hint for optimization purposes
xref:optimization_hints[optimization_hints].
```

Pennycook commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,602 @@
+_Effects_: Stores single element `in` to `out_iter` by using the `g` group
+object to identify memory location as `out_iter` + `g.get_local_linear_id()`
+
+`Properties` argument is reserved for future revisions of this extention and is
+ignored now.
+Default value is empty `sycl::ext::oneapi::experimental::empty_properties_t`
+May be used in future for setting boundary values or limiting numbers of work
+items.
```
Same question as above.  Can we use `contiguous` here?

Pennycook reviewed (commented):
I spotted one minor potential issue, but apart from this and the `full_group` discussion, I think we're good.

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+Example:
+`group_load(g, input, output_span, data_placement_blocked);`
+
+=== Optimization Hints
```
I think we should not use the word "hint" when describing these properties.  In other extensions, we use the word "hint" in cases when the property will not change the semantics of the program.  Even if you set a "hint" property wrong, the program will still run correctly.  The program may run slower when the hint is wrong, but it will never break.

We use the word "assertion" when the property could change the correctness of the program.  With an "assertion", the programmer is guaranteeing that something is true, and the program may run incorrectly if the property is set wrong.

Therefore, I'd suggest changing this section title to:

```suggestion
=== Optimization Properties
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+
+=== Optimization Hints
+
+anchor:optimization_hints[]
```
```suggestion
anchor:optimization_properties[]
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+As `InputIteratorT` and `OutputIteratorT` are permitted to be random access
+iterators they are not guaranteed to be contiguous.
+The following property is introduced to be used
+as a hint to the implementation for optimization purposes:
```
```suggestion
as an assertion to the implementation that can improve performance:
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+
+struct contiguous_memory_key {
+  using value_t =
+      property_value<contiguous_memory_hint>;
```
```suggestion
      property_value<contiguous_memory_key>;
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+      property_value<contiguous_memory_hint>;
+};
+
+inline constexpr contiguous_memory_hint::value_t contiguous_memory;
```
```suggestion
inline constexpr contiguous_memory_key::value_t contiguous_memory;
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+
+==== Groups partitioning
+
+The following property can be used as a hint that
```
```suggestion
The following property can be used as an assertion that
```

gmlueck commented on sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store.asciidoc:
```diff
@@ -0,0 +1,596 @@
+in striped case:
+`out_iter[g.get_local_linear_id() + g.get_local_linear_range() * i]` = `in[i];`
+for `i` between `0` and `ItemsPerWorkItem`.
+Properties may also provide a hint for optimization purposes
```
```suggestion
Properties may also provide assertions which can enable better optimizations
```

Or, it might read better to integrate the link like this:

```suggestion
Properties may also provide xref:optimization_properties[assertions] which can enable better optimizations.
```

And similar for the other place that have this sentence.

gmlueck reviewed (commented):
I think this looks good, but I have some comments about the use of the word "hint".  See below.

gmlueck reviewed (approved):
LGTM!

aelizaro commented:
@Pennycook, can we merge it now?

Pennycook commented:
> @Pennycook, can we merge it now?

I think so!  @intel/llvm-gatekeepers, please merge.
